WIP - Show c/storage (Buildah/CRI-O) containers in ps

TomSweeneyRedHat opened the pull request:
WIP - Show c/storage (Buildah/CRI-O) containers in ps

Ready for review.  I still have some work to do with testing, varlink and documentation.  But thought I'd get this out for a look/feel kind of review at the moment.

The `podman ps --all` command will now show containers that are under the control of other c/storage container systems and the new `ps --external` option will show only containers that are
in c/storage but are not controlled by libpod.

In the below examples, the '*working-container' entries were created by Buildah.

```
podman ps -a
CONTAINER ID  IMAGE                             COMMAND  CREATED       STATUS                   PORTS  NAMES
9257ef8c786c  docker.io/library/busybox:latest  ls /etc  8 hours ago   Exited (0) 8 hours ago          gifted_jang
d302c81856da  docker.io/library/busybox:latest  NA       30 hours ago  NA                              busybox-working-container
7a5a7b099d33  localhost/tom:latest              ls -alF  30 hours ago  Exited (0) 30 hours ago         hopeful_hellman
01d601fca090  localhost/tom:latest              ls -alf  30 hours ago  Exited (1) 30 hours ago         determined_panini
ee58f429ff26  localhost/tom:latest              NA       33 hours ago  NA                              alpine-working-container

podman ps --external
CONTAINER ID  IMAGE                             COMMAND  CREATED       STATUS  PORTS  NAMES
d302c81856da  docker.io/library/busybox:latest  NA       30 hours ago  NA             busybox-working-container
ee58f429ff26  localhost/tom:latest              NA       33 hours ago  NA             alpine-working-container

```


Signed-off-by: TomSweeneyRedHat <tsweeney@redhat.com>

AkihiroSuda commented:
Can they be treated as if they belong to namespaces? eg. podman --namespace=crio ps

TomSweeneyRedHat commented:
bot, retest this please.

rhatdan commented on libpod/runtime_ctr.go:
```diff
@@ -667,3 +676,58 @@ func (r *Runtime) GetLatestContainer() (*Container, error) {
 }
+
+//GetStorageContainers returns a list of containers from containers/storage that
+//are not currently known to Podmman.
```
Podman typo

TomSweeneyRedHat replied:
ty and fixed.

rhatdan commented on libpod/runtime_ctr.go:
```diff
@@ -667,3 +676,58 @@ func (r *Runtime) GetLatestContainer() (*Container, error) {
+
+//GetStorageContainers returns a list of containers from containers/storage that
+//are not currently known to Podmman.
+func (r *Runtime) GetStorageContainers(ctrs []*Container) ([]*Container, error) {
```
Horrible Function name.  GetNonLibpodContainers? GetExternalContainers?

rhatdan commented on pkg/adapter/runtime.go:
```diff
@@ -94,6 +99,14 @@ func (r *LocalRuntime) GetRWImages() ([]*ContainerImage, error) {
 }
 
+func (r *LocalRuntime) getImage(imageID string) (*image.Image, error) {
+	image, err := r.Runtime.ImageRuntime().GetImage(imageID)
```
Just return r.Runtime.ImageRuntime().GetImage(imageID)

TomSweeneyRedHat replied:
Actually looking at it further, I don't need this function and don't call it.  Backing this change out.

rhatdan commented on pkg/adapter/runtime_remote.go:
```diff
@@ -593,7 +593,7 @@ func (r *LocalRuntime) RemoveVolume(ctx context.Context, v *libpod.Volume, force
 // containers matching all filters are returned
-func (r *LocalRuntime) GetContainers(filters ...libpod.ContainerFilter) ([]*libpod.Container, error) {
+func (r *LocalRuntime) GetContainers(getOnlyStorage bool, filters ...libpod.ContainerFilter) ([]*libpod.Container, error) {
 	return nil, define.ErrNotImplemented
```
Why is this not implemented?

rhatdan commented:
I hate the name of the option.  I would figure --storage would be to return all containers in storage and
--storage=false would return only libpod containers.

Can we change it to --external 

podman ps -a --storage external 

@AkihiroSuda has a good idea, what the user would really want is something like

--namespace buildah or --namespace podman --namespace CRI-O, but all we can identify right now is whether or not the container was created via libpod or not.  Since Buildah and CRI-O do not use LIBPOD yet, we have no good way to identify where the storage containers came from.

mheon commented:
We already have a `--namespace` flag, though it's exclusively a libpod feature right now

rhatdan commented:
--namespace undefined? --namespace external?
Perhaps this would give me what I want.

podman ps --all
  Gives all containers in storage
podman ps --all --namespace=podman
  Gives all containers created by podman
podman ps --all --namespace=external
  Gives all containers not created with libpod
If CRI-O eventually grows libpod support
podman ps --all --namespace=cri-o
  Would give only cri-o containers.

mheon commented:
`--namespace` accepts arbitrary strings. It's used for Libpod namespaces, which were added with the intention of namespacing off CRI-O containers once we merged into CRI-O. We can't really repurpose it here as such.

mheon commented:
Also, I've just taken a brief look at the PR, and I'm pretty concerned that the logic here is leaking into libpod.

Libpod should never see a container not in the DB. We added separate functions for dealing with c/storage containers in `runtime_cstorage.go`; going through there prevents us from having to convince Libpod proper to deal with containers that don't exist in the DB. The cstorage functions don't give particularly much information, but there isn't much we can say about a container in the storage library if it doesn't exist in our DB - trying to populate `ps` as normal won't work, 90% of the fields can't be filled.

mheon commented on libpod/runtime_ctr.go:
```diff
@@ -667,3 +676,58 @@ func (r *Runtime) GetLatestContainer() (*Container, error) {
+			continue
+		}
+		newCtr := new(Container)
+		newCtr.config = new(ContainerConfig)
```
I would not expect building containers like this to work; soe of this, like lock allocation, has some definite undesirable side-effects.

We should confine the storage logic to cmd/podman, and ideally keep it separate from the rest of `ps`. If a storage container hits Libpod, I do not expect good things to happen.

TomSweeneyRedHat commented:
I've moved the external container list building logic from libpod into the podman ps logic.  I still need to test further, especially the remote stuff, but thought I'd put this up for review.  I've also renamed the option to `--external` and changed the underlying function.

mheon commented:
We should discuss this more on Monday - I still don't think that returning a `*libpod.Container` is viable here. We're going to unavoidably leak locks, for example.

I'd honestly prefer if we did not reuse the display logic from the rest of `ps` when displaying storage containers. We don't know enough about them to meaningfully populate all the fields in `ps`... But this is worth a more detailed discussion. Postscrum, maybe.

TomSweeneyRedHat commented:
Postscrum chat sounds good to me.

FWIW, the main reason I'm stuffing a storage container into a libpod container object is so that it can be shown in the `ps --all` and go through any filtering that a user asks to be done.  Assuming we keep this approach, is there something I should do to deallocate the locks at the end of the run?  I assumed, maybe badly, that they'd just go away at the end of the run when the in memory objects got cleaned up.

The other approach would be to just forget about showing the storage containers with the --all command.  But I think that's helpful for some one looking for one of these containers after they can't remove an image that has a storage container and not a libpod container still attached.

FWIW, I think I've found the remote compilation issue.  I'd to update a struct in `vi ./cmd/podman/shared/container.go`.  'make podman-remote' worked in my sandbox, does the CI do other compiling/linking?  It failed as I'd not included `External` to the PsOpts struct in that file.

mheon commented:
I think removing display in --all sounds fine; I'd prefer not to
intermingle containers from libpod and not from libpod. Would like to get a
consensus with Dan and Brent though.

On Sun, Sep 22, 2019, 13:38 Tom Sweeney <notifications@github.com> wrote:

> Postscrum chat sounds good to me.
>
> FWIW, the main reason I'm stuffing a storage container into a libpod
> container object is so that it can be shown in the ps --all and go
> through any filtering that a user asks to be done. Assuming we keep this
> approach, is there something I should do to deallocate the locks at the end
> of the run? I assumed, maybe badly, that they'd just go away at the end of
> the run when the in memory objects got cleaned up.
>
> The other approach would be to just forget about showing the storage
> containers with the --all command. But I think that's helpful for some one
> looking for one of these containers after they can't remove an image that
> has a storage container and not a libpod container still attached.
>
> FWIW, I think I've found the remote compilation issue. I'd to update a
> struct in vi ./cmd/podman/shared/container.go. 'make podman-remote'
> worked in my sandbox, does the CI do other compiling/linking? It failed as
> I'd not included External to the PsOpts struct in that file.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/containers/libpod/pull/4053?email_source=notifications&email_token=AB3AOCDQJZ4AO262ILR4JGLQK6UPVA5CNFSM4IXQKN32YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7JLCXI#issuecomment-533901661>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB3AOCATTEVXC4DKTJU5IMDQK6UPVANCNFSM4IXQKN3Q>
> .
>

rhatdan commented:
@TomSweeneyRedHat Needs a rebase.  Is this something we should talk about at the f2f?

TomSweeneyRedHat commented:
I think we're all set talk wise, I just need to carve out some time and get back to this.

rhatdan commented:
@TomSweeneyRedHat I guess this has bubbled back up to importance, at least do a rebase, Please.

rhatdan commented:
@TomSweeneyRedHat Could you rebase this one, and we can start the conversation again.

TomSweeneyRedHat commented:
I've just rebased this, no changes yet, so please don't review.

rhatdan commented:
Replacing with https://github.com/containers/podman/pull/7290
